c/tskit/tables.c

```diff
@@ -8184,6 +8205,230 @@ tsk_table_collection_subset(
     return ret;
 }
 
+static int
+tsk_check_subset_equality(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *other_node_mapping, tsk_size_t num_shared_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, i;
+    tsk_id_t *self_nodes = NULL;
+    tsk_id_t *other_nodes = NULL;
+    tsk_table_collection_t self_copy;
+    tsk_table_collection_t other_copy;
+
+    memset(&self_copy, 0, sizeof(self_copy));
+    memset(&other_copy, 0, sizeof(other_copy));
+    self_nodes = malloc(num_shared_nodes * sizeof(*self_nodes));
+    other_nodes = malloc(num_shared_nodes * sizeof(*other_nodes));
+    if (self_nodes == NULL || other_nodes == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    i = 0;
+    for (k = 0; k < (tsk_id_t) other->nodes.num_rows; k++) {
+        if (other_node_mapping[k] != TSK_NULL) {
+            self_nodes[i] = other_node_mapping[k];
+            other_nodes[i] = k;
+            i++;
+        }
+    }
+
+    // TODO: strict sort before checking equality
+    ret = tsk_table_collection_copy(self, &self_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_copy(other, &other_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&other_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&self_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&self_copy, self_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&other_copy, other_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    if (!tsk_table_collection_equals(&self_copy, &other_copy)) {
+        ret = TSK_ERR_UNION_DIFF_HISTORIES;
+        goto out;
+    }
+
+out:
+    tsk_table_collection_free(&self_copy);
+    tsk_table_collection_free(&other_copy);
+    tsk_safe_free(other_nodes);
+    tsk_safe_free(self_nodes);
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_union(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *other_node_mapping, tsk_flags_t options)
+{
+    int ret = 0;
+    tsk_id_t k, i, new_parent, new_child;
+    tsk_size_t num_shared_nodes = 0;
+    tsk_edge_t edge;
+    tsk_mutation_t mut;
+    tsk_site_t site;
+    tsk_id_t *node_map = NULL;
+    tsk_id_t *individual_map = NULL;
+    tsk_id_t *population_map = NULL;
+    tsk_id_t *site_map = NULL;
+    bool add_populations = !(options & TSK_UNION_NO_ADD_POP);
+    bool check_shared_portion = !(options & TSK_UNION_NO_CHECK_SHARED);
+
```

Comment:
I've just realised that we should be doing an integrity check before running this because we're chasing lots of references and not checking if they are valid. Something like

ret = tsk_table_collection_check_integrity(self, 0);
if (ret != 0) {
    goto out;
}
ret = tsk_table_collection_check_integrity(other, 0);
if (ret != 0) {
    goto out;
}
Then, to set it, just set one of the population references to -2 or something in the node table.

We should also do this for subset().